In [1]:
def convert_rgb_hsl(r,g,b):
    R = r / 255
    G = g / 255
    B = b / 255
    minimum = min([R,G,B])
    maximum = max([R,G,B])
    l = ((minimum+maximum)/2)*100
    if l>50:
        s = ((maximum-minimum)/(2.0-maximum-minimum))*100
    else:
        s = ((maximum-minimum)/(maximum+minimum))*100

    if maximum == R:
        h = ((G-B)/(maximum-minimum))*60
    elif maximum == G:
        h = (2.0 + ((B-R)/(maximum-minimum)))*60
    else:
        h = (4.0 + ((R-G)/(maximum-minimum)))*60
    return (h,s,l)

In [2]:
def Read_Image(filename):
    im = Image.open(filename)
    #im = im.crop((550,100,650,500))
    pixel_values = list(im.getdata())
    #im.show()
    #print(pixel_values)
    dict_colors = {}
    for pixel in pixel_values:
        if pixel not in dict_colors:
            dict_colors[pixel] = 1
        else:
            dict_colors[pixel] += 1
    blue_red = [0,0] #store number of blues (0) and reds (1)
    for key in dict_colors.keys():
        if  dict_colors[key] > 20 and key != (255,255,255,255) and key != (0,0,0,255)and key[0]!=key[2]:
            (h,s,l) = convert_rgb_hsl(key[0], key[1], key[2])
            if l> 20:
                if h>(170) and h<(300):#color is blue
                    #print(key)
                    #print('h = '+ str(h))
                    #print('s = '+ str(s))
                    #print('l = '+ str(l))
                    blue_red[0] +=1
                else:#color is red
                    blue_red[1] +=1
    return blue_red
        

In [3]:
def getFeatures(file_list):
    feature_list =[[]]*len(file_list)
    counter = 0
    for file in file_list:
        example = Read_Image(file)
        feature_list[counter]=example
        counter+=1
    return feature_list

In [10]:
import os
import numpy as np
from PIL import Image
file_list = os.listdir(os.curdir)
file_list = file_list[1:(len(file_list)-1)]
x = getFeatures(file_list)

In [11]:
y = ['bad','okay','okay','okay', 'good','good','good','okay','good','good','good','bad', 'good','bad','bad','bad','bad', 'good','good', 'okay', 'bad','good', 'okay', 'good', 'okay', 'good', 'okay', 'okay', 'okay', 'okay', 'good', 'good', 'good', 'bad',  'good', 'good','okay','okay', 'okay', 'good', 'okay', 'bad',  'bad',  'good', 'okay', 'okay', 'bad',  'bad',  'good', 'okay', 'good', 'good', 'okay', 'okay', 'okay', 'good', 'okay', 'good', 'good', 'good', 'good', 'good', 'good', 'good', 'good', 'okay', 'okay', 'good', 'good', 'good', 'okay', 'good','okay','good','bad','okay', 'good','good','good','good','good', 'good','good','okay', 'okay', 'good','okay','okay', 'okay','good','good','okay','okay','good','okay', 'good', 'good', 'good', 'good', 'okay','good', 'good', 'good', 'okay', 'okay', 'okay', 'good', 'good', 'good', 'good', 'okay', 'okay', 'good', 'good', 'bad', 'bad', 'okay']

In [50]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2)
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=8)
knn.fit(x_train, y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=8, p=2,
           weights='uniform')

In [51]:
from sklearn import metrics
p = knn.predict(x_test)
metrics.accuracy_score(p, y_test)

0.83333333333333337

Accuracy of KNN (k=8) = 0.83333333333333337

One reason for innaccuracy is that classifications used for training are done bby hand and may be inconsistent.

In [62]:
from sklearn import svm
clf = svm.SVC(gamma=.007, C=5)
clf.fit(x_train, y_train)  
p2 = clf.predict(x_test)
metrics.accuracy_score(p2, y_test)

0.75

Accuracy of Using SVM classification = 0.75

# Report summary: 

Although K Nearest Neighbor was more accurate, it is not consistent when the training set varies. Sometime k=4 or k=3 are more accurate if the training set is different. 

The accuracy of the SVM also varied when the training set was varied. 